src/argprocessing.cpp

```diff
@@ -39,7 +39,7 @@ struct ArgumentProcessingState
   bool found_S_opt = false;
   bool found_pch = false;
   bool found_fpch_preprocess = false;
-  bool found_color_diagnostics = false;
+  signed char found_color_diagnostics = 0; // 0=auto, -1=no, 1=yes
```

Comment:
I actually did initially write it like that, but there are two reasons I backed away from that:

* Without specifying an underlying type, an `enum class` can (and probably will) be represented as an `int`. That would be a waste of space, both because of the three wasted bytes inside the type and because of the alignment padding it would introduce to `struct ArgumentProcessingState`. Of course, it's easy enough to tell the compiler that the underlying type should be `uint8_t`, so this is a non-issue.
* The bigger reason is that comparisons against zero (`x == 0`, `x != 0`, `x < 0`, `x > 0`, etc.) are smaller in code (and faster to execute on some architectures) than comparisons against non-zero immediate values. I would rather check `found_color_diagnostics > 0` than `color_diagnostics == ColorDiagnostics::always` because the former can be implemented using fewer and shorter instructions.

A side point: if I were to use an enumerated type here, I would position `automatic` as the first enumerator (having value zero) so that the structure can be zero-initialized en masse and there isn't a odd-ball "one" byte in the middle.

Reply:
@jrosdahl: I [did](https://github.com/ccache/ccache/pull/596/commits/9f24dc661e608f1bb472134d66097e2c257c35e9) as you requested in a way that's almost as efficient as what I had and is potentially clearer in the source code.

Reply:
Thanks for explaining your reasoning.

My view is that optimizations like these of course are good when they matter, but since ArgumentProcessingState is just created once and the field in question is written to a handful of times this feels like a severe case of premature optimization. Your update looks OK.

Reply:
I'd agree with all of that. I'd just prefer be told to be less efficient than to be told to be more efficient. :wink: